### PR TITLE
Wait for input writer to complete before calling OnConnectionClosed

### PR DESCRIPTION
**Pull request by halter73**
#2565

### CODE REVIEW

**Comment by halter73** [comment]
@davidfowl please take a look.

File: src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs [hunk 1]
```diff
@@ -69,22 +70,43 @@ public Task OnConnectionAsync(ConnectionContext connectionContext)
...
             {
-                ((HttpConnection)state).OnConnectionClosed(error);
-            },
-            connection);
+                Connection = connection;
+            }
 
-            return processingTask;
+            public HttpConnection Connection;
+            public TaskCompletionSource<object> CompletedTcs = new TaskCompletionSource<object>();
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs]: Do you want to run these continuations inline?

File: src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs [hunk 2]
```diff
@@ -69,22 +70,43 @@ public Task OnConnectionAsync(ConnectionContext connectionContext)
...
+            await inputCompletionState.CompletedTcs.Task;
+            await outputCompletionState.CompletedTcs.Task;
+
+            connection.OnConnectionClosed();
 
-            connectionContext.Transport.Output.OnReaderCompleted((error, state) =>
+            await processingTask;
+        }
+
+        private static void CompletedCallback(Exception error, object state)
```
Comment (pakrym) [inline src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs]: Move this into `PipeCompletionState`

Comment (pakrym) [reply src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs]: And only expose `Task` from it.

File: src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs (hunk 1 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs]: The pipe completion callbacks already get dispatched by default. This is closer to calling Abort and OnConnectionClosed directly in the callback like before.

File: src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs (hunk 2 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs]: But now I'll have to cast twice 😢

**Pull request merged**